### PR TITLE
Rework error handling of asn1_do_lock method.

### PR DESCRIPTION
**Pull request by FdaSilvaYY**
Related to exchange in PR #1030
Cleanup of  some FIXME tags.

### CODE REVIEW

File: crypto/asn1/tasn_fre.c [hunk 1]
```diff
@@ -89,7 +89,8 @@ static void asn1_item_embed_free(ASN1_VALUE **pval, const ASN1_ITEM *it,
 
     case ASN1_ITYPE_NDEF_SEQUENCE:
     case ASN1_ITYPE_SEQUENCE:
-        if (asn1_do_lock(pval, -1, it) > 0)
+        count = asn1_do_lock(pval, -1, it);
```
Comment (richsalz) [inline crypto/asn1/tasn_fre.c]: what's the point of adding the 'count' variable?  and is changing the test the right thing?

Comment (FdaSilvaYY) [inline crypto/asn1/tasn_fre.c]: I add it to not reuse 'i'.
The test is changed, to continue the free-ing code only when ref-count is 0 only, 
and not on error ( < 0 ) ...

Comment (richsalz) [inline crypto/asn1/tasn_fre.c]: I mean, why use a local variable at all?

File: crypto/asn1/tasn_utl.c [hunk 2]
```diff
@@ -46,13 +46,14 @@ int asn1_set_choice_selector(ASN1_VALUE **pval, int value,
 }
 
 /*
- * Do reference counting. The value 'op' decides what to do. if it is +1
- * then the count is incremented. If op is 0 count is set to 1. If op is -1
- * count is decremented and the return value is the current reference count
- * or 0 if no reference count exists.
- * FIXME: return and manage any error from inside this method
+ * Do atomic reference counting. The value 'op' decides what to do.
```
Comment (mattcaswell) [inline crypto/asn1/tasn_utl.c]: Whitespace on the end of this line

File: crypto/asn1/tasn_utl.c [hunk 3]
```diff
@@ -46,13 +46,14 @@ int asn1_set_choice_selector(ASN1_VALUE **pval, int value,
...
 /*
- * Do reference counting. The value 'op' decides what to do. if it is +1
- * then the count is incremented. If op is 0 count is set to 1. If op is -1
- * count is decremented and the return value is the current reference count
- * or 0 if no reference count exists.
- * FIXME: return and manage any error from inside this method
+ * Do atomic reference counting. The value 'op' decides what to do. 
+ * If it is +1 then the count is incremented.
+ * If |op| is 0, lock is initialised and count is set to 1.
+ * If |op| is -1, count is decremented and the return value is the current
```
Comment (mattcaswell) [inline crypto/asn1/tasn_utl.c]: Whitespace

File: include/openssl/asn1.h [hunk 4]
```diff
@@ -885,6 +885,7 @@ void ERR_load_ASN1_strings(void);
 # define ASN1_F_ASN1_D2I_READ_BIO                         107
 # define ASN1_F_ASN1_DIGEST                               184
 # define ASN1_F_ASN1_DO_ADB                               110
+# define ASN1_F_ASN1_DO_LOCK                              233
```
Comment (mattcaswell) [inline include/openssl/asn1.h]: Whitespace

**Comment by mattcaswell** [comment]
I'm happy with this, subject to the extra whitespace on the end of lines that I noted in comments being removed. I wonder whether this code could be simplified even further. AFAICT there are only two callers of asn1_do_lock...one which sets `op` to 0, and one with `op` as -1...but never `op` as +1!!!??? Maybe I'm just missing something - we'd probably need to involve Steve Henson in any further simplification - but I don't see that as part of the scope of this MR.

+1

File: crypto/asn1/tasn_utl.c [hunk 5]
```diff
@@ -67,21 +68,24 @@ int asn1_do_lock(ASN1_VALUE **pval, int op, const ASN1_ITEM *it)
     lck = offset2ptr(*pval, aux->ref_offset);
     lock = offset2ptr(*pval, aux->ref_lock);
     if (op == 0) {
-        *lck = 1;
+        *lck = 1;   /* Set reference count to 1 */
```
Comment (richsalz) [inline crypto/asn1/tasn_utl.c]: remove the comment.

**Comment by FdaSilvaYY** [comment]
@mattcaswell : I'm do the same analysis as you. I bet that this code is call by some d2i method whose output is a X509/X509_REQ/X509_CRL object with his lock fields to up-to-date.

**Comment by mattcaswell** [comment]
@richsalz is this ok now?

**Comment by richsalz** [comment]
LGTM +1  You merge or me?

**Comment by mattcaswell** [comment]
Merged. Thanks.

**Pull request closed without merging**